### PR TITLE
Pico extensibility support for the OCI SDK

### PR DESCRIPTION
**Pull request by trentjeff**
Addresses issue #6406 and #6499.

### CODE REVIEW

**Comment by trentjeff** [comment]
The README files on the PR should explain the feature.

This PR fully addresses https://github.com/helidon-io/helidon/issues/6406. However, please review my comments on https://github.com/helidon-io/helidon/issues/6499 since there is some drift on this PR with regard to the original requirements.

File: integrations/oci/sdk/processor/src/main/resources/io/helidon/integrations/oci/sdk/processor/no.dot.exceptions
Comment (trentjeff) [reply integrations/oci/sdk/processor/src/main/resources/io/helidon/integrations/oci/sdk/processor/no.dot.exceptions:2]: If you look at the template (here is one small snippet):
```
        this.builderProvider = (InjectionPointProvider<{{simpleclassname}}Client{{dot}}Builder>) builderProvider;
```
The OCI SDK does not follow the pattern for these API services:
```
com.oracle.bmc.streaming.Stream
com.oracle.bmc.streaming.StreamAsync
```
For these there is "no dot" in the inner class design.

File: integrations/oci/sdk/runtime/pom.xml
Comment (trentjeff) [reply integrations/oci/sdk/runtime/pom.xml]: done and pushed

**Review by tomas-langer — APPROVED** [review]

File: integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java
Comment (ljnelson) [inline integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java:55]: Or https://docs.oracle.com/en-us/iaas/tools/java/3.17.0/com/oracle/bmc/auth/AbstractFederationClientAuthenticationDetailsProviderBuilder.html#METADATA_SERVICE_BASE_URL

File: integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java
Comment (ljnelson) [inline integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java:58]: Or https://docs.oracle.com/en-us/iaas/tools/java/3.17.0/com/oracle/bmc/auth/AbstractFederationClientAuthenticationDetailsProviderBuilder.html#FALLBACK_METADATA_SERVICE_URL

File: integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java
Comment (ljnelson) [inline integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java:73]: Why have an explicit singular form?

"auth" is a bad name; I don't like it in the CDI extension either

File: integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java
Comment (ljnelson) [inline integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java:77]: This class does not attempt these strategies. It is abstract. I think you mean `OciAuthenticationDetailsProvider` (which is not itself an ADP)?

No need to say "when one is called for"

File: integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java
Comment (ljnelson) [inline integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java:76]: "strategies" -> "strategy descriptors" if you're staying with the terminology you direct the reader to

File: integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java
Comment (ljnelson) [inline integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java:124]: "It is expected to be passed with" -> "It must be"

File: integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java
Comment (ljnelson) [inline integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java:136]: `fileConfigIsPresent`?

File: integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java
Comment (ljnelson) [inline integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java:322]: This is a weird name.

File: integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java
Comment (ljnelson) [inline integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java:317]: I don't understand this.

File: integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java
Comment (ljnelson) [inline integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java:319]: "there is" -> "there are"...not sure frankly what this sentence is saying

File: integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java
Comment (ljnelson) [inline integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java:331]: See above

File: integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java
Comment (ljnelson) [inline integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java:336]: Are these methods basically saying, "The user has supplied a strategy descriptor of `foo`. Let's see if `foo` could work."?  If so, that is not clear to me. Do they need to be `public` and exposed and therefore managed from version to version, or are they implementation details that could be hidden away somewhere?

**Review by ljnelson — COMMENTED** [review]
I've left a boatload of comments as you can see.

Since this PR has already been approved I will let its approval status stay.  I will leave it up to you what, if any, of my comments you want to address; I certainly don't have all the answers. I think you could let 'er rip and then the issues could be addressed in an ongoing fashion. I hope my comments are clear; I've tried to place myself in the shoes of a prospective user as I read through the 56 files in this PR and tried to capture my reactions as I encountered things.

As you requested, I focused mostly on the `AuthenticationDetailsProvider` stuff. I think it's generally OK, but I confess I had a little trouble figuring out the configuration and why and how it differs from the current strategies. As long as the user can continue to "do nothing" and the "right" ADP is chosen for her application's runtime environment, that's the most important thing, and, if I'm reading right, that is indeed the case.

At a higher level, I found the language, naming and conceptual structure of all this to be quite loose and in some cases actively surprising, and tried to comment on all the places that were strange or confusing or surprising to me.  Some of this was confined to this PR, and some of it is in the overall Helidon Injection core concepts and APIs.

Finally, by not requesting changes on this PR I don't want to be understood to be approving all of Helidon Injection; I think eventually its language, concepts and structures need a hard, careful, thorough, rigorous look, much more documentation, and perhaps even a specification describing exactly what must be true of various implementations of various interfaces and constructs.

Anyway I hope my review is useful and thanks for the work.

File: integrations/oci/sdk/README.md
Comment (trentjeff) [reply integrations/oci/sdk/README.md:11]: will improve in a follow-up PR.  Keeping this (and other comments below) open as a TODO reminder.

File: integrations/oci/sdk/README.md
Comment (trentjeff) [reply integrations/oci/sdk/README.md:13]: Yes, the two modules are listed in the readme, and DI is defined in the top most overview on Pico.  What is missing is a link to the top level Pico readme, so keeping this open as a note to add that on the next PR.

File: integrations/oci/sdk/README.md
Comment (trentjeff) [reply integrations/oci/sdk/README.md:15]: ok, next PR.

File: integrations/oci/sdk/README.md
Comment (trentjeff) [reply integrations/oci/sdk/README.md:20]: Suggestion for a better choice?  It's anything that is "part of the MP flavor", based upon MicroProfile + CDI...

File: integrations/oci/sdk/README.md
Comment (trentjeff) [reply integrations/oci/sdk/README.md:39]: this is needed.

File: integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java
Comment (trentjeff) [reply integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java:322]: Suggest a better name? It is used to determine the availability of the OCI File-based profile scheme.

File: integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java
Comment (trentjeff) [reply integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java:317]: i'm not sure what else I could say to make it clearer.

File: integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java
Comment (trentjeff) [reply integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java:319]: next PR

File: integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java
Comment (trentjeff) [reply integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java:331]: Same answer - what can make it clearer?

File: integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java
Comment (trentjeff) [reply integrations/oci/sdk/runtime/src/main/java/io/helidon/integrations/oci/sdk/runtime/OciConfigBean.java:336]: Not really, no. This method (and the other one above) basically ensures that all of the requisite config attributes are indeed present so that it could suffice to be used for the Simple OCI Auth Provider.

**Pull request merged**
